store/tx_manager.go

```diff
@@ -17,44 +20,51 @@ const defaultGasLimit uint64 = 500000
 // the local Config for the application, and the database.
 type TxManager struct {
 	*EthClient
-	KeyStore *KeyStore
-	Config   Config
-	ORM      *models.ORM
+	keyStore      *KeyStore
+	config        Config
+	orm           *models.ORM
+	activeAccount *ActiveAccount
 }
 
 // CreateTx signs and sends a transaction to the Ethereum blockchain.
 func (txm *TxManager) CreateTx(to common.Address, data []byte) (*models.Tx, error) {
-	account, err := txm.KeyStore.GetAccount()
-	if err != nil {
-		return nil, err
-	}
-	nonce, err := txm.GetNonce(account.Address)
-	if err != nil {
-		return nil, err
-	}
-	tx, err := txm.ORM.CreateTx(
-		account.Address,
-		nonce,
-		to,
-		data,
-		big.NewInt(0),
-		defaultGasLimit,
-	)
-	if err != nil {
-		return nil, err
+	if txm.activeAccount == nil {
+		return nil, errors.New("Must activate an account before creating a transaction")
 	}
+
 	blkNum, err := txm.GetBlockNumber()
 	if err != nil {
 		return nil, err
 	}
 
-	gasPrice := &txm.Config.EthGasPriceDefault
-	_, err = txm.createAttempt(tx, gasPrice, blkNum)
-	if err != nil {
-		return tx, err
-	}
+	var tx *models.Tx
+	txm.activeAccount.GetAndIncrementNonce(func(nonce uint64) error {
+		tx, err = txm.orm.CreateTx(
+			txm.activeAccount.Address,
+			nonce,
+			to,
+			data,
+			big.NewInt(0),
+			defaultGasLimit,
+		)
+		if err != nil {
+			return err
+		}
+
+		gasPrice := txm.config.EthGasPriceDefault
+		var txa *models.TxAttempt
+		txa, err = txm.createAttempt(tx, &gasPrice, blkNum)
+		if err != nil {
+			txm.orm.DeleteStruct(tx)
```

Comment:
Looks like we delete structs from the database because we can't access wrapping this all in a database transaction. Understandable given our limitations with Bolt, but maybe we could refactor `createAttempt` to additionally receive a "Saver", at which point we could pass in a bolt's database transaction, or pass in an regular ORM instance to save. One edge case that makes me want a transaction: what if one call to `DeleteStruct` errs, and the other doesn't?

Reply:
In your edge case, he deliberately fails silently and keeps it moving. Two ramifications of this: 1. we can continue if there isn't something to delete, 2. we fail to delete something and keep going.

The spots where we would need to pass a `saver` would be:
1. txm.orm.CreateTx
2. txm.createAttempt
3. txm.ORM.AddAttempt (inside createAttempt)

It's a lot of passing, but I still prefer a tx rollback to the manual delete. How about this idea, we implement our own minimal version of nestable transactions? We're definitely going to need this again. Intercept calls to `orm.Begin(true)` to return a `TxORM` that delegates relevant calls to the stored `dbtx`.

Super pseudo code:

```golang
type TxORM {
        *ORM
        dbtx storm.Node // https://github.com/asdine/storm/blob/fa32729ddab40b4511820cf9ba364002a6992706/node.go#L34
}

func (orm *ORM) Begin(writable bool) (*TxORM,error) {
        dbtx, err := orm.Begin(true)
        return &TxORM{orm, dbtx}, err
}

func (txorm *TxORM) Begin(writable bool) (*TxORM,error) {
        return txorm
}

// redirects to dbtx
func (txorm *TxORM) Save ...
func (txorm *TxORM) Rollback ...
func (txorm *TxORM) Commit ...
```

We'll need to create a small interface to support polymorphism but perhaps this could be a cleaner solution than passing saver.

Reply:
Spiked on the approach suggested by @dimroc but decided to pull this out as a separate story because it also had a deadlock. https://www.pivotaltracker.com/story/show/157104995

Reply:
Cool idea, but might be worth waiting until we run into this again. Might be worth seeing other ways contexts where this problem shows up before coming up with a general answer.